### PR TITLE
Port process-type

### LINKED ISSUE #737: Port process-type
```c
DEFUN ("process-type", Fprocess_type, Sprocess_type, 1, 1, 0,
       doc: /* Return the connection type of PROCESS.
The value is either the symbol `real', `network', `serial', or `pipe'.
PROCESS may be a process, a buffer, the name of a process or buffer, or
nil, indicating the current buffer's process.  */)
  (Lisp_Object process)
{
  Lisp_Object proc;
  proc = get_process (process);
  return XPROCESS (proc)->type;
}
```

### PR DESCRIPTION
**Pull request by leemachin**
Closes #737 

Left out a test for serial processes because I'm not sure how to do that without depending on something like socat, but there's enough there to prove it works :)

### CODE REVIEW

File: rust_src/src/process.rs [hunk 1]
```diff
@@ -263,6 +268,15 @@ pub fn process_status(process: LispObject) -> LispObject {
     status
 }
 
+/// Return the connection type of PROCESS.
+/// The value is either the symbol `real', `network', `serial', or `pipe'.
+/// PROCESS may be a process, a buffer, the name of a process or buffer, or
+/// nil, indicating the current buffer's process.
+#[lisp_fn]
+pub fn process_type(process: LispProcessRef) -> LispObject {
+    process.process_type()
```
Comment (nitishch) [inline rust_src/src/process.rs]: The argument to `process-type` can either be a process/buffer/string/nil. When you define the argument as of type `LispProcessRef`, it accepts only process argument.

Comment (leemachin) [reply rust_src/src/process.rs]: Had a feeling it wasn't so simple ;) thanks!

Comment (shaleh) [reply rust_src/src/process.rs]: Indeed. You need the call to `get_process` or its equivalent.

**Comment by leemachin** [comment]
You might guess I'm fairly fresh to Rust so if there's a more idiomatic approach to what I've done here I will gladly accept the feedback. Other than that, it should behave as expected now.

**Comment by shaleh** [comment]
Do not stress being new to Rust. We are happy to help people learn.

File: rust_src/src/process.rs [hunk 2]
```diff
@@ -263,6 +268,32 @@ pub fn process_status(process: LispObject) -> LispObject {
     status
 }
 
+/// Return the connection type of PROCESS.
+/// The value is either the symbol `real', `network', `serial', or `pipe'.
+/// PROCESS may be a process, a buffer, the name of a process or buffer, or
+/// nil, indicating the current buffer's process.
+#[lisp_fn]
+pub fn process_type(process: LispObject) -> LispObject {
+    let mut p;
```
Comment (shaleh) [inline rust_src/src/process.rs]: Whenever you mark a variable as mutable you should ask why. If the algorithm really requires updates that is expected. In the code below you can use the fact that if/else is also an expression which returns a value.

File: rust_src/src/process.rs [hunk 3]
```diff
@@ -263,6 +268,32 @@ pub fn process_status(process: LispObject) -> LispObject {
...
+    };
+
+    if p.is_nil() {
+        return p;
+    }
+
+    let p_ref = p.as_process_or_error();
+
+    p_ref.process_type()
+}
```
Comment (shaleh) [inline rust_src/src/process.rs]:
```
let process = if process.is_string() {
    let p = get_process(process);
    if p.is_nil() {
        get_buffer_process(process)
    } else {
        unsafe { cget_process(process) }  // no more to_raw
   }
};

if process.is_nil() {
    process
} else {
    let p_ref = process.as_process_or_error();
    p_ref.process_type()
}
```

Comment (shaleh) [reply rust_src/src/process.rs]: Overwriting the function argument with the transformed version of it is also a common idiom.

Comment (shaleh) [reply rust_src/src/process.rs]: The original code uses a single function called `get_process`. Why not port it using the code above as a beginning? I see there is a `get_process` in this Rust file already. You can rename it to `get_process_lisp` and set the name to be `get_process` in the lisp_fn macro so it still works in C.

**Comment by leemachin** [comment]
Makes sense 👍 saw a few places where this logic exists (with very subtle variations in some places, e.g. handling of processes when the string can refer to either a process name or a buffer name) so it makes sense to abstract it out and do a bit of cleanup here.

File: rust_src/src/process.rs (hunk 3 above)
Comment (leemachin) [reply rust_src/src/process.rs]: Great :) I thought the `mut` was weird, just distracted myself a bit with the nested conditionals.

File: rust_src/src/process.rs [hunk 4]
```diff
@@ -84,15 +89,36 @@ impl LispProcessRef {
...
+/// current buffer.
+pub fn get_process(process: LispObject) -> LispProcessRef {
+    let process = if process.is_string() {
+        let p = lisp_get_process(process);
+        if p.is_nil() {
+            get_buffer_process(process)
+        } else {
+            p
+        }
+    } else if process.is_nil() {
```
Comment (shaleh) [inline rust_src/src/process.rs]: That should not be an else. It should be a stand alone if.

Comment (leemachin) [reply rust_src/src/process.rs]: Good spot.

**Comment by leemachin** [comment]
Back to work today so progress on finishing this off might be a little slower, but I've more or less grasped what needs to be done.

**Comment by shaleh** [comment]
No worries, it will be waiting for you. :smile:

You just need to be ready to rebase.

File: rust_src/src/process.rs [hunk 5]
```diff
@@ -84,15 +89,52 @@ impl LispProcessRef {
...
+    }
+
+    if proc_or_buf.is_buffer() {
+        let proc = get_buffer_process(proc_or_buf);
+        if proc.is_nil() {
+            panic!("Buffer {} has no process", name.as_string_or_error())
+        }
+        proc
+    } else {
+        // check process
```
Comment (shaleh) [inline rust_src/src/process.rs]: Should this not return the process?

Comment (leemachin) [reply rust_src/src/process.rs]: There's some stuff in the original C function that does some extra validation (there are bits and pieces to see if the process is actually running or not, it seems). I just need to figure out what exactly that is.

**Comment by shaleh** [comment]
brotzeit has picked this up in a new PR.

**Pull request closed without merging**